hicommonwealth/commonwealth
# Session keys: SIWE, one-way sync for signed actions to p2p operations log

## Link to Issue
Closes: #3302 

## Description of Changes
- [x] Changes Ethereum login from EIP712 to SIWE, for better security usability
- [x] Adds a GET /api/oplog endpoint, which queries for recent signed actions, and exports recent actions
- [ ] ~~Adds a POST /api/oplog endpoint, which allows new action+session tuples to be submitted and executed against CW.~~
- [x] Adds a canvas-node package to the Commonwealth runtime, which can be run using `yarn start` or `yarn dev` to start a Canvas node that syncs against the CW API endpoint

We currently use a single SQL query with a UNION statement to provide GET /api/oplog. Later versions should use a Postgres trigger to maintain another table or materialized view with actions and sessions.

## Test Plan
- Start CW server in dev mode
- Go to /packages/canvas-node, run `yarn dev` or `yarn start`
    - yarn dev uses the locally installed `canvas`
    - yarn start uses the version of canvas specified inside packages/canvas-node
- Interact with all chains, by creating a thread, comment, and reactions
    - Ethereum
    - Cosmos
        - Vanilla Cosmos (e.g. Osmosis)
        - Terra
        - Evmos
        - Injective
    - Polkadot
    - Solana
    - NEAR
- Observe that all actions have synchronized successfully to peer

## Deployment Plan

Nothing special to do here. We can set up the production Canvas node at any point, after this goes in.

## Other Considerations

- Next PR: Inbound sync should be straightforward, since the signature formats are all synced up now.
- Next PR: Peering policies / restrictions on libp2p connections.
